compiler/src/iree/compiler/Dialect/Flow/Transforms/RaiseSpecialOps.cpp

```diff
@@ -76,6 +76,197 @@ std::optional<Value> matchATransposeBMatmul(linalg::LinalgOp matmulOp) {
   return std::nullopt;
 }
 
+/// Matches a linalg.generic operation reading data from a tensor `source` using
+/// tensor.extract, and raises the `source` tensor to an input of the linalg
+/// operation.
+static LogicalResult raiseTensorExtractToInput(linalg::GenericOp linalgOp,
+                                               RewriterBase &rewriter) {
+  if (!linalgOp.hasTensorSemantics())
+    return failure();
+  if (!isElementwise(linalgOp))
+    return failure();
+
+  // Find a tensor.extract op in the linalgOp body.
+  tensor::ExtractOp extractOp;
+  linalgOp.walk([&](tensor::ExtractOp op) {
+    if (extractOp)
+      return WalkResult::interrupt();
+    extractOp = op;
+    return WalkResult::advance();
+  });
+
+  if (!extractOp)
+    return failure();
+
+  // Raise the tensor.extract op to an input.
+  SmallVector<AffineExpr> exprs;
+  for (Value indexValue : extractOp.getIndices()) {
+    // For raising, the indexing value must be one of the following:
+    //    1. A constant value.
+    //    2. A linalg.index.
+
+    // 1. Indexing value is a constant.
+    APInt constantIndex;
+    if (matchPattern(indexValue, m_ConstantInt(&constantIndex))) {
+      exprs.push_back(getAffineConstantExpr(constantIndex.getLimitedValue(),
```

Comment:
I feel like tensor.extract in general is a bigger evil than having constants in an affine map. I would always prefer a tensor.extract to be raised to an input, even if it produces constants in an affine map. So while, I agree it's bad, it's a lesser evil for me.

Also, another reason I match for constants in the raising to view pattern is a lot of inputs use a constant for dimensions of size 1, instead of getting the dimension from linalg.index. So we have to support constants in AffineMap in these raising patterns.

Reply:
Ok, my concern is more when the raising to `linalg.generic` happens and the conversion to `tensor.extract_slice` doesnt happen. How about we split the difference (and I do like the intermediate `linalg.generic`) : For a gather-like op convert it to the intermediate `linalg.generic` and return it (i.e. dont replace uses of the gather-like op with the `linalg.generic`). Use that to immediately convert the `linalg.generic` to `tensor.extract_slice`. If the second conversion succeeds, replace the original gather-like op with the `tensor.extract_slice`. If it doesnt, just leave the original gather-like op as is (the intermediate generated will get DCEed or you can explicitly delete it as well).